---
title: 'ggvita: Visualization for Tree Alignment'
author: "Jianrong Philip Yang, Meng Yuan, Xiaolong Cao"
date: '2017-10-10'
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

R package, ggvita, is developed for visualization of tree alignment results. These results connect genotypes with phenotypes during a binary-tree development, from single cell ( usually zygote ) to multiple cells. Pro. Jianrong Philip Yang designed an algorithm to calculate the similarity between trees, in Sun Yat-sen University, Guang Zhou. This algorithm can dectect tree similar phenotypes at the sub tree scale. When we connect both genotypes and phenotypes of livings, we can observe some interesting phenomena and find out the biology mechanisms behind them. stat_EPIC function can add expression levels of genes from EPIC (http://epic.gs.washington.edu/) on the branches of trees alignment results.


**The results should be obtained through DELTA software FIRSTLY.**

## Local alignment 

* Read in the result from DELTA. 
```
library(ggvita)

# alm_file <- "the address of ggvita data -- fun.alm "

# alml_file <- "the address of ggvita data -- fun.alml "

# cost <- "the address of cost file --- cost.tsv"

# alml_list <- readal(outfile= alml_file, fileS=alm_file, fileT=alml_file,cost=cost)

```
ggvita. This will only show the basic tree alignment result with structures of treeS and treeT.

```{r}
library(ggvita)
library(ggtree)
library(ggplot2)


p <- ggvita(ggvita::alml_list, 3)
p
```

*  "%++%" is a ggvita linking synax to add the ggvita objects, similar to the system of ggplot2. We can view the leaves tissue by adding tippoint layer.

```{r}
 p %++% geom_tippoint(aes(fill=I(tip.fill)),size=1,shape=21,color="NA")

```

* stat_epic. This will show the tree alignment result with both genotype and phenotype of treeS and treeT.
This is an example of PHA-4 gene expression.

```{r}

#expr_file <- normalizePath("./CD20061215_pha4I2L_11.csv")

# stat_epic_example <- stat_epic(p,expr_file = expr_file,size=1,mc.cores = 1) (Use ggvita::stat_epic_example to show.)
pp <- 
p %++%  
  ggvita::stat_epic_example %++% 
  geom_tippoint(aes(fill=I(tip.fill)),size=1,shape=21,color="NA") %++% 
  scale_color_gradientn(colors =rainbow(7)[1:7],
                        values = c(0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8)/3
                        )

pp


```



* stat_prune. This will show the prune nodes number in tree alignment results of treeS and treeT.

```{r}

pp %++%
  stat_prune(p,size=1,color="blue")
  
```




**Have fun!**  


Contact with authors via [yuanm26@mail2.sysu.edu.cn](yuanm26@mail2.sysu.edu.cn)

